### PR TITLE
ISSUE-196-custom-file-storage-path --- provide options to set relative paths for storage of managed strawberryfield files and digital object files (json)

### PR DESCRIPTION
**Pull request by patdunlavey**
See #196 

This PR provides configuration options on the persistent file storage settings page to set relative paths for storage of the strawberryfield uploaded files (default none) and the digital object files (json) (default "dostorage").

### CODE REVIEW

**Comment by patdunlavey** [comment]
@DiegoPino let me know your thoughts. Thanks!

File: config/schema/strawberryfield.schema.yml
Comment (DiegoPino) [inline config/schema/strawberryfield.schema.yml:82]: Ok, do we need both? if we add here the file_path then the alter hook needs to be adapted again? Just wondering since the ISSUE was about the Object File Path. But happy to deal with both (we need an alter too in the form)

File: src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php
Comment (DiegoPino) [inline src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php:103]: This looks good. What if you make `dostorage` a class constant?

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [inline src/Form/FilePersisterServiceSettingsForm.php:53]: Let's add a warning here. That says "changing this path after your ingests won't modify existing  Files"

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [inline src/Form/FilePersisterServiceSettingsForm.php:54]: Wonder if we should better go for NULL instead of empties? And then change the logic of the setting? Not sure really

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [inline src/Form/FilePersisterServiceSettingsForm.php:320]: Lovely! But I think for this case we need to make sure these paths are sanitized. Also S3 has size limits. Look at the method where I refactored the hook. I have a few notes there about this

File: src/StrawberryfieldFilePersisterService.php
Comment (DiegoPino) [inline src/StrawberryfieldFilePersisterService.php:237]: Ok, this needs to be added to the refactor too

**Comment by DiegoPino** [comment]
@patdunlavey thanks for the pull. I added a few questions really more than code reviews. Looks good, i would mostly like to do some more sanitization on the inputs and make sure there is no way a user can corrupt storage via this forms. This also means I need to refactor my #199 to match the new paths.

File: config/schema/strawberryfield.schema.yml
Comment (patdunlavey) [reply config/schema/strawberryfield.schema.yml:82]: Yes, this is not strictly necessary since we can modify the path through hook_strawberryfield_file_destination_alter(). However, it seems to me that being able to set the path would be desired pretty often, and requiring a user to implement a hook just to set this to a different path seems a bit onerous, particularly as they will see that we do give them the configuration option for the digital object file storage.
I don't feel strongly about this though, Happy to take it out if you think that makes more sense.

File: src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php
Comment (patdunlavey) [reply src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php]: Done.

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php]: Ok, I added this note to both fields. Tell me if that's not correct!

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php:56]: My thinking was that the user input for this field is trimmed, exploded on slashes, etc - all operations that expect a string.

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php:324]: So checkPlain and limit the path to 128 characters? Ok to just set the field character limit?

File: src/StrawberryfieldFilePersisterService.php
Comment (patdunlavey) [reply src/StrawberryfieldFilePersisterService.php:237]: I'm thinking let's get your refactor merged first, and then resolve this branch it. Does that make sense?

Comment (DiegoPino) [reply src/StrawberryfieldFilePersisterService.php:237]: Ok, let's roll with that. Will wait for your review on that pull before merging it. Thx

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [reply src/Form/FilePersisterServiceSettingsForm.php:324]: I think we need to make folder validation method. Should not have spaces. Should have a controlled list of characters (like no utf8 strawberries).  Here is a list of what needs to be checked. I agree we are also allowing local storage, e.g what @giancarlobi does, but still not sure I want to have on a local unix filesystem something that fights a 1986 unix spec too much. See https://www.ezs3.com/public/232.cfm

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [reply src/Form/FilePersisterServiceSettingsForm.php:56]: Ok, let's roll with an empty string

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [reply src/Form/FilePersisterServiceSettingsForm.php]: Thanks. Correct!

File: config/schema/strawberryfield.schema.yml
Comment (DiegoPino) [reply config/schema/strawberryfield.schema.yml:82]: Ok, let's keep it. I understand how this can be useful for vanilla implementations!

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php:324]: Ok, I'll work on a validation method.

**Comment by patdunlavey** [comment]
Update: #199 has been merged. I need to address the file path sanitation issues (in the next day or so) and then will ask for further review.

**Comment by patdunlavey** [comment]
I've added validation for the two file path fields - both ajax and upon submit. It's probably overkill, but it's working pretty good for me. The one thing that had me scratching my head for a good while was how to validate the system file path (public and private schemes). As I commented in the code, neither \Drupal::service('path.validator')->isValid() nor \Drupal::service('stream_wrapper_manager')->isValidUri() work --- at all! Well, I probably don't understand how they're supposed to work. I ended up using FileSystem::prepareDirectory() to test if the directory path provided could be written to. Can you review that part in particular, @DiegoPino ? There is almost certainly a better way to do it!

**Comment by patdunlavey** [comment]
Hi @DiegoPino. With all the discussion of the CSV exporter today I don't want to lose sight of this issue being the only one we know of that is actually blocking us from starting to ingest content into our production instance. Any chance you could take a bit of time to give this a look?

**Comment by DiegoPino** [comment]
@patdunlavey I can tomorrow. Did you adapt the code so it merges clean with the last hook change? This will break my open https://github.com/esmero/strawberryfield/pull/189 so give me at least 24 hours, hands totally full. Will check back when ready. thx

**Comment by patdunlavey** [comment]
@DiegoPino yes it merges clean with the last hook change (that change has been merged to 1.0.0-RC3 and this PR is against that). I'll take a look at #189. Maybe I can create a combined merge branch? Or if that branch is ready to merge, maybe you could do that and then I can reconcile again with this branch?

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [inline src/Form/FilePersisterServiceSettingsForm.php:87]: The `exec` is a bit confusing. Better remove since there is no execution at all

File: src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php
Comment (DiegoPino) [inline src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php:106]: Like this

File: src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php
Comment (DiegoPino) [inline src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php:26]: really like this. Could you add a few "_" to make it more legible? Sorry. It really helps

File: config/schema/strawberryfield.schema.yml
Comment (DiegoPino) [inline config/schema/strawberryfield.schema.yml:92]: good!

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:35]: This is nice. Could we make this one more explicit? VALID_FILESYSTEM_PATH_PATTERN?

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:42]: Same here (also super nice)  VALID_S3_PREFIX_PATTERN? (since in S3 paths are normally known as prefixes)

**Review by DiegoPino — CHANGES_REQUESTED** [review]
@patdunlavey this is pretty much in good shape but still needs a bit more work. There is a bit of redundancy in the code (validation methods can be a single) one and S3 v/s any other are handled differently when it does not need to. Some CONST (thanks for that) need spacing (_) to be readable and a few minor things. I can help you tomorrow to get those done and we can merge. Thanks again for working on this.

File: src/StrawberryfieldUtilityService.php
Comment (patdunlavey) [reply src/StrawberryfieldUtilityService.php:373]: Thanks for the dependency injection correction - this symfony stuff is still very foggy to me!
I'm with you that creating a directory and then deleting it seems dangerous. I was hoping you could point me to another way to determine if the proposed destination path is writable!
I'll leave the two-step logic as-is unless and until we can think of something more elegant.

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php:87]: Yep, a lazy copy-paste mistake.

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php:201]: I'm nervous about removing the regex tests because I'm not 100% sure I like the invasive method of determining writability. Am I just being silly?

Comment (DiegoPino) [reply src/Form/FilePersisterServiceSettingsForm.php:201]: Oh, do not remove them. What I say here is that you can deal with any wrapper in the same way. and to be honest, to allow portability between filesystem, the MOST restrictive regex should be the one that is applied to both (so the S3). That makes also your code simpler @patdunlavey . That way moving from local filesystem to the cloud is seamless. If not you have to update all your ADOS!

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php]: Fixed.

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php]: Fixed.

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php]: See latest commit. Fairly major refactoring of a lot of this stuff since we are changing to using the same validation rule for S3 and local file systems.

File: src/StrawberryfieldUtilityService.php
Comment (patdunlavey) [reply src/StrawberryfieldUtilityService.php]: Changes committed. Please re-review.

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php]: Ok, switched to using the S3 prefix pattern for all. Please re-review!

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php]: Fixed.

File: src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php
Comment (patdunlavey) [reply src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php]: Better now, I hope!

File: src/StrawberryfieldUtilityService.php
Comment (patdunlavey) [reply src/StrawberryfieldUtilityService.php]: Better now, I hope.

File: src/StrawberryfieldUtilityService.php
Comment (patdunlavey) [reply src/StrawberryfieldUtilityService.php]: Better now I hope.

**Comment by patdunlavey** [comment]
I think I have addressed all your review comments. The main thing was refactoring to only use one validation method (S3). Got rid of a fair bit of redundant code. Made the error message even more wordy! Ready for your re-review when you get a chance. Thanks so much @DiegoPino , you are amazing!

**Comment by DiegoPino** [comment]
@patdunlavey amazing! Will give it a spin at 2:30 and will let you know. Thanks!!

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (DiegoPino) [inline src/Form/FilePersisterServiceSettingsForm.php:429]: @patdunlavey this looks much better. I will leave this at it is but will refactor tonight a bit. Why?

We have this here:
 $scheme_options = OcflHelper::getVisibleStreamWrappers();

If a scheme is not in that list, then it is actually invalid and we need to validate as "false". No worries. It works the way you wrote but I feel its good we reuse code.

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:104]: @patdunlavey I see a lot of DCS changes here. Some super good (thx) other a bit crazy and overly strict. I will also give this file a second pass after the merge to restore some of lost read-ability (but still keeping it DCS)

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:191]: nice! gosh.

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:201]: @patdunlavey like this one. Totally crazy how PHP Storm reformatted this

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:263]: good!

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:372]: @patdunlavey You trimming this somewhere right? Like "       " becomes "" ?

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:376]: Again, we have the OCFL toolbox I wrote to fetch all known schemes that are active and can be used to writing. Will refactor

File: src/StrawberryfieldUtilityService.php
Comment (patdunlavey) [reply src/StrawberryfieldUtilityService.php:104]: Sorry about that. I didn't notice until I had already committed and pushed. If you want, I can reverse these.

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [inline src/StrawberryfieldUtilityService.php:411]: @patdunlavey better (for readability) to set $valid to FALSE at the beginning. That way you always return$valid

File: src/Form/FilePersisterServiceSettingsForm.php
Comment (patdunlavey) [reply src/Form/FilePersisterServiceSettingsForm.php:429]: Makes sense. Thanks!

**Comment by DiegoPino** [comment]
@patdunlavey so I added a few comments. I will merge at it is tonight but will do a small refactor. The DCS pass on the strawberryfield utilities is a bit crazy and there are a few pieces of code that can be simplified by reusing (see same form) existing methods (visible stream wrappers). Thanks. Please do not merge until I refactor after merging. Will mention you. Hugs!

File: src/StrawberryfieldUtilityService.php
Comment (patdunlavey) [reply src/StrawberryfieldUtilityService.php:376]: Ok, thanks for your patience with a D8+ noob!

**Comment by patdunlavey** [comment]
Hugs backatcha!

File: src/StrawberryfieldUtilityService.php
Comment (DiegoPino) [reply src/StrawberryfieldUtilityService.php:104]: No worries. Php Storm does a lot to conform to the 80 lines limit but DCS does not apply that rule for functions. I can do that

**Review by DiegoPino — APPROVED** [review]
ready to merge. Small refactor coming later to night! Thanks @patdunlavey on your work on this!

**Pull request merged**
